src/src/PluginStructs/P037_data_struct.h

```diff
@@ -0,0 +1,117 @@
+#ifndef PLUGINSTRUCTS_P037_DATA_STRUCT_H
+#define PLUGINSTRUCTS_P037_DATA_STRUCT_H
+
+#include "../../_Plugin_Helper.h"
+#include "../../ESPEasy_common.h"
+#include "../CustomBuild/StorageLayout.h"
+
+#ifdef USES_P037
+
+# define PLUGIN_037_DEBUG     // Additional debugging information
+
+# define P037_MAPPING_SUPPORT // Enable Value mapping support
+# define P037_FILTER_SUPPORT  // Enable filtering support
+# define P037_JSON_SUPPORT    // Enable Json support
+
+// #ifdef LIMIT_BUILD_SIZE       // Leave out the fancy stuff is available flash is tight
+// #  ifdef P037_MAPPING_SUPPORT
+//      #undef P037_MAPPING_SUPPORT
+// #  endif
+// #  ifdef P037_FILTER_SUPPORT
+//      #undef P037_FILTER_SUPPORT
+// #  endif
+// #  ifdef P037_JSON_SUPPORT
+//      #undef P037_JSON_SUPPORT
+// #  endif
+// #endif
+
+# define P037_MAX_MAPPINGS  25
+# define P037_MAX_FILTERS   25
+# define P037_EXTRA_VALUES  5
+
+# define P037_OPERAND_COUNT 2
+# define P037_OPERAND_LIST  F("=%")
+
+# define P037_FILTER_COUNT  2 // 3rd option not implemented yet
+# define P037_FILTER_LIST   F("=-:")
+
+// Data structure
+struct P037_data_struct : public PluginTaskData_base
+{
+  P037_data_struct(taskIndex_t taskIndex);
+  ~P037_data_struct();
+
+  bool   webform_load(
+#ifdef P037_MAPPING_SUPPORT
+                      bool mappingEnabled
+#endif
+#if defined(P037_MAPPING_SUPPORT) && defined(P037_FILTER_SUPPORT)
+                      ,
+#endif
+#ifdef P037_FILTER_SUPPORT
+                      bool filterEnabled
+#endif
+#if (defined(P037_MAPPING_SUPPORT) || defined(P037_FILTER_SUPPORT)) && defined(P037_JSON_SUPPORT)
+                      ,
+#endif
+#ifdef P037_JSON_SUPPORT
+                      bool jsonEnabled
+#endif
+                     );
+  bool   webform_save(
+#ifdef P037_FILTER_SUPPORT
+                      bool filterEnabled
+#endif
+#if defined(P037_FILTER_SUPPORT) && defined(P037_JSON_SUPPORT)
+                      ,
+#endif
+#ifdef P037_JSON_SUPPORT
+                      bool jsonEnabled
+#endif
+                     );
+#ifdef P037_MAPPING_SUPPORT
+  String mapValue(String input);
+  void   logMapValue(String input, String result);
+#endif // P037_MAPPING_SUPPORT
+#ifdef P037_FILTER_SUPPORT
+  bool   hasFilters();
+  bool   checkFilters(String key, String value);
+#ifdef PLUGIN_037_DEBUG
+  void   logFilterValue(String text, String key, String value, String match);
+#endif // PLUGIN_037_DEBUG
+#endif // P037_FILTER_SUPPORT
+
+  // The settings structures
+  // The stuff we want to save between settings
+  struct tP037_StoredSettings_struct {
+    char deviceTemplate[VARS_PER_TASK][41];		// variable for saving the subscription topics, leave as first element for backward compatibility

```

Comment:
Please see how it is done in `P036_data_struct::loadDisplayLines` where the display lines are loaded.
`LoadCustomTaskSettings` also has a version to load an array of `String` so you only allocate the memory that's being used instead of large blocks of data.
Saving it can also be done from an array of `String`, so you can still keep the old format.

Problem with those large arrays as you're now using is that they need to be allocated as a large continuous block which may become impossible if the heap memory gets fragmented.

Reply:
We might need to have a 1:1 chat about that (discord/slack/whatever)

Reply:
Prima :)

Reply:
Are changes still required here? I reduced the memory/storage usage down by half of max. available storage (when mappings or filters enabled and using only 4 filters), so that looks efficient enough, IMHO.
I have looked at the P036 example mentioned, but can't see where I could apply that mechanism to this structure 🤔